# TCC PUC-Minas Gastão Jucá Filho <br>
Turma Puc-Minas Virtual - nov2019 <br>


#  Atividades:
> 1. Resumir os passos dos últimos 03 (três) notebooks. 
> 2. Montar um tabelão Olist com os campos que serão trabalhados na classificação:
>> Partir das 08 (oito) tabelas da Olist Dataset, previamente preparadas. 
> 3. Abusar dos merges.
> 4. Documentar tudo com linguagem Markdown.

In [1]:
# Eis o DataSet da Olist original com que iniciamos o trabalho

from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "As oito tabelas iniciais.PNG")

####  Trabalhamos todas as 08 (oito) tabelas de modo a deixá-las normalizadas de jeito que um banco de dados relacional requer.
> Abaixo resumimos as operações efetuadas em cada tabela, para após, criarmos nosso tabelão.
>> A operação é necessária para chegarmos ao DataFrame que conterá o "tabelão" de trabalho.

##  1 - Importação de bibliotecas necessárias

In [2]:
import pandas as pd
import numpy as np
pd.options.display.float_format = '{:,.2f}'.format
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
#pd.options.display.max_columns = 100


## 2 - Importação das 08 (oito) tabelas 

In [3]:
reviews = pd.read_csv("olist_order_reviews_dataset.csv", sep = ',')

In [4]:
orders = pd.read_csv('olist_orders_dataset.csv', sep=',')

In [5]:
itens = pd.read_csv("olist_order_items_dataset.csv", sep = ',')

In [6]:
products = pd.read_csv("olist_products_dataset.csv", sep = ',')

In [7]:
customers = pd.read_csv("olist_customers_dataset.csv", sep = ',')

In [8]:
geolocation = pd.read_csv("olist_geolocation_dataset.csv", sep = ',')

In [9]:
payments = pd.read_csv("olist_order_payments_dataset.csv", sep = ',')

In [10]:
sellers = pd.read_csv("olist_sellers_dataset.csv", sep = ',')

In [11]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [12]:
geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037           -23.55           -46.64   
1                         1046           -23.55           -46.64   
2                         1046           -23.55           -46.64   
3                         1041           -23.54           -46.64   
4                         1035           -23.54           -46.64   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

## 3 - Execução resumida das operações realizadas nos prévios 03 (três) notebooks
> Transformação dos dados desde os CSV originais até o ponto de se iniciar a montagem do tabelão, objeto deste notebook.

### 3.1 - Tabela <font color=red>"orders"

#### A tabela <font color=red>"orders" não precisou de limpeza alguma.

### 3.2 - Tabela <font color=red>"reviews"

In [13]:
# Deletamos os duplicados do campo 'review_id'
review_id_sem_repetir = reviews.drop_duplicates (subset=['review_id'])

In [14]:
# Depois, deletamos os duplicados do campo 'order_id'
reviews_normalizada = review_id_sem_repetir.drop_duplicates (subset=['order_id']) 

In [15]:
reviews_normalizada.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 98926 entries, 0 to 99999
Data columns (total 7 columns):
review_id                  98926 non-null object
order_id                   98926 non-null object
review_score               98926 non-null int64
review_comment_title       11659 non-null object
review_comment_message     41343 non-null object
review_creation_date       98926 non-null object
review_answer_timestamp    98926 non-null object
dtypes: int64(1), object(6)
memory usage: 6.0+ MB


### 3.3 - Tabela <font color=red>"itens"

In [16]:
# Deletamos duplicatas do campo "order_item_id"
itens_sem_repetir = itens.drop_duplicates (subset=['order_id']) # Será mantida a 1ª ocorrência.

In [17]:
# situação APÓS da deleção
Image (url = "tres tabelas pos delecao.bmp")

#### Apesar de a nova tabela <font color=red>"itens_sem_repetir"</font> falhar em informar sobre itens, ainda é útil para informar qual o produto e qual o fornecedor estão atrelados a cada pedido.
> No entanto, vamos desprezar em uma futura operação de 'Join' as colunas restantes, pois:  
>> Os preços do produto e do frete influenciariam somente em uma análise de 'Churn' que não é o nosso caso.  
>> A data de envio também é desnecessária, uma vez que temos datas diversas na tabela <font color=red>"orders"</font>.

### 3.4 - Tabela <font color=red>"products"

#### A tabela <font color=red>"products" não precisou de limpeza alguma.

#### <font color=magenta>RESUMO</font>. Não há a descrição do nome do produto, somente o seu código. Não podemos ver qual produto é, apenas a sua categoria. Há 74 categorias para os mais de 32 mil produtos.
> 'product_id' é a chave primária da tabela <font color=red>"products"</font>

> Não há o que limpar, a tabela <font color=red>"products"</font> está pronta para uso.

### 3.5 - Tabela <font color=red>"payments"

##### <font color=magenta>RESOLUÇÃO</font> 
> Criaremos os seguintes campos (para serem associados a cada 'order_id'):
>> 1. Qtde de boleto  
2. Qtde de credit_card  
3. Qtde de debit_card  
4. Qtde de not_defined  
5. Qtde de voucher  
6. Qtde de Installments  
7. Soma de Payment_value

> E desprezaremos os campos 'payment_sequential', 'payment_type', 'payment_installments' e 'payment_value'

#### 3.5.1 - PREPARAÇÃO da implementação da criação de colunas novas na tabela <font color=red>"payments" </font>

                                          1. Qtde de 'boleto'

In [18]:
# seleciona 'order_id' cujo 'payment_type' = 'boleto'
payments_somente_boleto = payments.loc[payments['payment_type']=='boleto']

In [19]:
# Mudar o nome da coluna para 'qtde_boleto'
payments_somente_boleto.columns = ['order_id', 'payment_sequential', 'qtde_boleto', 'payment_installments', 'payment_value']

In [20]:
# substituir 'boleto' pelo algarismo '1'.
payments_somente_boleto = payments_somente_boleto.apply(lambda x: x.replace('boleto', 1))

###### DataFrame <font color=red>"payments_somente_boleto"</font> pronto para join

                                          2. Qtde de 'credit_card'

In [21]:
# seleciona 'order_id' cujo 'payment_type' = 'credit_card'
payments_somente_credit_card = payments.loc[payments['payment_type']=='credit_card']

In [22]:
# Contagem de 'credit_card' de cada 'order_id'
qtde_credit_card = payments_somente_credit_card.groupby('order_id')['payment_type'].count()

df_qtde_credit_card = pd.DataFrame(qtde_credit_card)

In [23]:
# Tira 'order_id' da condição de índice.
df_qtde_credit_card_index_resetado = df_qtde_credit_card.reset_index()

In [24]:
# Mudar o nome da coluna para 'qtde_credit_card'
df_qtde_credit_card_index_resetado.columns = ['order_id', 'qtde_credit_card']

###### DataFrame <font color=red>"df_qtde_credit_card_index_resetado"</font> pronto para join

                                          3. Qtde de 'debit_card'

In [25]:
# seleciona 'order_id' cujo 'payment_type' = 'boleto'
payments_somente_debit_card = payments.loc[payments['payment_type']=='debit_card']

In [26]:
# Contagem de 'debit_card' de cada 'order_id'
qtde_debit_card = payments_somente_debit_card.groupby('order_id')['payment_type'].count()

df_qtde_debit_card = pd.DataFrame(qtde_debit_card)

In [27]:
# Tira 'order_id' da condição de índice.
df_qtde_debit_card_index_resetado = df_qtde_debit_card.reset_index()

In [28]:
# Mudar o nome da coluna para 'qtde_credit_card'
df_qtde_debit_card_index_resetado.columns = ['order_id', 'qtde_debit_card']

###### DataFrame <font color=red>"df_qtde_debit_card_index_resetado"</font> pronto para join

                                          4. Qtde de 'not_defined'

In [29]:
# seleciona 'order_id' cujo 'payment_type' = 'not_defined'
payments_somente_not_defined = payments.loc[payments['payment_type']=='not_defined']

In [30]:
# Mudar o nome da coluna para 'qtde_not_defined'
payments_somente_not_defined.columns = ['order_id', 'payment_sequential', 'qtde_not_defined', 'payment_installments', 'payment_value']

In [31]:
# substituir 'not_defined' pelo algarismo '1'.
payments_somente_not_defined = payments_somente_not_defined.apply(lambda x: x.replace('not_defined', 1))

###### DataFrame <font color=red>"payments_somente_not_defined"</font> pronto para join

                                          5. Qtde de 'voucher'

In [32]:
# seleciona 'order_id' cujo 'payment_type' = 'voucher'
payments_somente_voucher = payments.loc[payments['payment_type']=='voucher']

In [33]:
# Contagem de 'voucher' de cada 'order_id'
qtde_voucher = payments_somente_voucher.groupby('order_id')['payment_type'].count()

df_qtde_voucher = pd.DataFrame(qtde_voucher)

In [34]:
# Tira 'order_id' da condição de índice.
df_qtde_voucher_index_resetado = df_qtde_voucher.reset_index()

In [35]:
# Mudar o nome da coluna para 'qtde_credit_card'
df_qtde_voucher_index_resetado.columns = ['order_id', 'qtde_voucher']

###### DataFrame <font color=red>"df_qtde_voucher_index_resetado"</font> pronto para join

                                          6. Soma de 'payment_value'

In [36]:
# Contagem de 'payment_installments' de cada 'order_id'
soma_payment_value = payments.groupby('order_id')['payment_value'].sum()

df_soma_payment_value = pd.DataFrame(soma_payment_value)

In [37]:
# Tira 'order_id' da condição de índice.
df_soma_payment_value_index_resetado = df_soma_payment_value.reset_index()

###### DataFrame <font color=red>"df_soma_payment_value_index_resetado"</font> pronto para join

                                          7. Qtde de 'installments'

In [38]:
# Contagem de 'payment_installments' de cada 'order_id'
# Usaremos a função 'max()'
qtde_installments = payments.groupby('order_id')['payment_installments'].max()

df_qtde_installments = pd.DataFrame(qtde_installments)

In [39]:
# Tira 'order_id' da condição de índice.
df_qtde_installments_index_resetado = df_qtde_installments.reset_index()

###### DataFrame <font color=red>"df_qtde_installments_index_resetado"</font> pronto para join

                                 ------------------ x -------------------

#### 3.5.2 - IMPLEMENTAÇÃO da criação de colunas novas na tabela <font color=red>"payments" </font>

In [40]:
# deleção de todos os campos que não sejam 'order_id'
payments_somente_order = payments.drop(['payment_sequential', 'payment_type', 'payment_installments', 'payment_value'], axis=1)

###### Os DataFrames usados no join são:
> 1. payments_somente_boleto   
2. df_qtde_credit_card_index_resetado  
3. df_qtde_debit_card_index_resetado  
4. payments_somente_not_defined  
5. df_qtde_voucher_index_resetado  
6. df_soma_payment_value_index_resetado  
7. df_qtde_installments_index_resetado

                                     1.Merge para obter 'qtde_boleto'

In [41]:
# exemplo de merge  --->    m = pd.merge(tabela_1, tabela_2, how = 'left', on = 'Nome')
payments_boleto = pd.merge(payments_somente_order, payments_somente_boleto, how = 'left', on = 'order_id')

In [42]:
# Deleção dos campos não mais necessários
payments_boleto_limpo = payments_boleto.drop(['payment_sequential', 'payment_installments','payment_value'], axis=1)

                                     2.Merge para obter 'qtde_credit_card'

In [43]:
boleto_credit_card = pd.merge(payments_boleto_limpo, df_qtde_credit_card_index_resetado, how = 'left', on = 'order_id')

                                     3.Merge para obter 'qtde_debit_card'

In [44]:
boleto_credit_e_debit_card = pd.merge(boleto_credit_card, df_qtde_debit_card_index_resetado, how = 'left', on = 'order_id')

                                     4.Merge para obter 'qtde_not_defined'

In [45]:
boleto_credit_debit_NotDefined = pd.merge(boleto_credit_e_debit_card, payments_somente_not_defined, how = 'left', on = 'order_id')

In [46]:
# Deleção dos campos não mais necessários
boleto_credit_debit_NotDefined_limpo = boleto_credit_debit_NotDefined.drop(['payment_sequential', 'payment_installments','payment_value'], axis=1)

                                     5.Merge para obter 'qtde_voucher'

In [47]:
bol_cred_deb_Ndef_voucher = pd.merge(boleto_credit_debit_NotDefined_limpo, df_qtde_voucher_index_resetado, how = 'left', on = 'order_id')

                                     6.Merge para obter 'soma_payment'

In [48]:
bol_cred_deb_Ndef_vou_pay = pd.merge(bol_cred_deb_Ndef_voucher, df_soma_payment_value_index_resetado, how = 'left', on = 'order_id')

In [49]:
# renomeando a última coluna:
bol_cred_deb_Ndef_vou_pay.columns = ['order_id', 'qtde_boleto', 'qtde_credit_card', 'qtde_debit_card', 'qtde_not_defined', 'qtde_voucher', 'soma_payment']

                                     7.Merge para obter 'qtde_installments'

In [50]:
bol_cred_deb_Ndef_vou_pay_install = pd.merge(bol_cred_deb_Ndef_vou_pay, df_qtde_installments_index_resetado, how = 'left', on = 'order_id')

In [51]:
# renomeando a última coluna:
bol_cred_deb_Ndef_vou_pay_install.columns = ['order_id', 'qtde_boleto', 'qtde_credit_card', 'qtde_debit_card', 'qtde_not_defined', 'qtde_voucher', 'soma_payment', 'qtde_installments']

####  A nova tabela <font color=red>"payments"</font>:

> Se chamará <font color=red>"payments_normalizada"</font>

> Consistirá da tabela <font color=red>"bol_cred_deb_Ndef_vou_pay_install"</font> (recém obtida), sem duplicidade de
'order_id'

In [52]:
# Deleção dos duplicados de 'order_id' para torná-lo a chave primária da nova tabela
payments_normalizada = bol_cred_deb_Ndef_vou_pay_install.drop_duplicates (subset=['order_id'])

In [53]:
# Configuração atual APÓS a implementação da resolução.
Image (url = "cinco tabelas - payments_normalizada.bmp")

                                 ------------------ x -------------------

### 3.6 - Tabela <font color=red>"customers"

In [54]:
customers.head(1)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   

   customer_zip_code_prefix customer_city customer_state  
0                     14409        franca             SP

#### <font color=magenta>RESUMO</font>. Esses dois primeiros campos podem suscitar dúvidas, vamos esclarecer:
> <font color=green>'customer_id'</font> é a chave primária da tabela <font color=red>"customers"</font>
>> Também é único na tabela <font color=red>"orders"</font>, ou seja, cada compra tem seu 'customer_id'. Acho essa modelagem boboca, mas é assim que está.

> <font color=green>'Customer_unique_id'</font> é único para a cada pessoa. 
>> Se uma pessoa fizer duas compras na Olist, em cada uma, receberá um <font color=green>'customer_id'</font> diferente. Gostaram?

> Não há o que limpar, a tabela <font color=red>"customers"</font> está pronta para uso.

### 3.7 - Tabela <font color=red>"sellers"</font>

#### 3.7.1 - Adição do campo <font color=green>'business_segment'</font> à tabela <font color=red>"sellers"</font>
> A partir da tabela <font color=green>'closed_deals'</font>, do database OLIST MARKETING FUNNEL

In [55]:
# Leitura da tabela "closed deals"
closed_deals = pd.read_csv("olist_closed_deals_dataset.csv", sep = ',')

In [56]:
closed_deals.head(1)

mql_id                         seller_id  \
0  5420aad7fec3549a85876ba1c529bd84  2c43fb513632d29b3b58df74816f1b06   

                             sdr_id                             sr_id  \
0  a8387c01a09e99ce014107505b92388c  4ef15afb4b2723d8f3d81e51ec7afefe   

              won_date business_segment      lead_type lead_behaviour_profile  \
0  2018-02-26 19:58:54              pet  online_medium                    cat   

  has_company has_gtin average_stock business_type  \
0         NaN      NaN           NaN      reseller   

   declared_product_catalog_size  declared_monthly_revenue  
0                            nan                      0.00

In [57]:
closed_deals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
mql_id                           842 non-null object
seller_id                        842 non-null object
sdr_id                           842 non-null object
sr_id                            842 non-null object
won_date                         842 non-null object
business_segment                 841 non-null object
lead_type                        836 non-null object
lead_behaviour_profile           665 non-null object
has_company                      63 non-null object
has_gtin                         64 non-null object
average_stock                    66 non-null object
business_type                    832 non-null object
declared_product_catalog_size    69 non-null float64
declared_monthly_revenue         842 non-null float64
dtypes: float64(2), object(12)
memory usage: 92.2+ KB


In [58]:
# Por que só há 842 'seller_id' na tabela "closed_deals" (tabela da direita) ????? Deveria haver 3.095, como na da esquerda.
Image (url = "sellers e closed_deals.bmp")

In [59]:
# Fazendo merge de "sellers" com "closed_deal" para obter o campo 'business_segment'
sellers_com_bSegment = pd.merge(sellers, closed_deals, how = 'left', on = 'seller_id')

In [60]:
# Deleção dos campos não mais necessários
sellers_normalizado = sellers_com_bSegment.drop(['mql_id', 'sdr_id','sr_id', 'won_date', 'lead_type', 'lead_behaviour_profile', 'has_company', 'has_gtin', 'average_stock', 'business_type', 'declared_product_catalog_size', 'declared_monthly_revenue'], axis=1)

In [61]:
sellers_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3095 entries, 0 to 3094
Data columns (total 5 columns):
seller_id                 3095 non-null object
seller_zip_code_prefix    3095 non-null int64
seller_city               3095 non-null object
seller_state              3095 non-null object
business_segment          380 non-null object
dtypes: int64(1), object(4)
memory usage: 145.1+ KB


###### Somente 380 'business_segment' em 3.090 'seller_id'.
> É outra incongruência desse DataSet.
>> Provavelmente falha no carregamento de dados.

In [62]:
# Poderíamos relacionar 'product_category_name' (da tabela "products") à 'business_segment' (da tabela "sellers_normalizada")
Image (url = "sete tabelas com sellers_normalizada.bmp")

### 3.8 - Tabela <font color=red>"geolocation"</font>

In [63]:
# Eliminando repetidos na concatenação de todos os campos.
geolocation_sem_repetir = geolocation.drop_duplicates (subset=['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state'])

##### A chave primária é a concatenação de TODOS os campos da tabela.

##### A nova configuração a partir da qual montaremos o tabelão está abaixo.

In [64]:
Image (url = "Tabelas quase completas.PNG")

## 4 - Montagem do Tabelão

##### Partiremos da tabela <font color=red>"orders"</font>, indicada pelas setas, e pegaremos campos marcados pelos retângulos em <font color=green>"verde"</font>.

In [65]:
Image (url = "Campos iniciais para tabelao.PNG")

### 4.1 - Merge <font color=red>"orders"</font> com <font color=red>"reviews_normalizada"</font>
> Manter toda a tabela <font color=red>"orders"</font> e trazer somente a correspondência de <font color=red>"reviews_normalizada"</font>

In [66]:
orders_reviews_apos_merge = pd.merge(orders, reviews_normalizada, how = 'left', on = 'order_id')

In [67]:
orders_reviews_apos_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99441 entries, 0 to 99440
Data columns (total 14 columns):
order_id                         99441 non-null object
customer_id                      99441 non-null object
order_status                     99441 non-null object
order_purchase_timestamp         99441 non-null object
order_approved_at                99281 non-null object
order_delivered_carrier_date     97658 non-null object
order_delivered_customer_date    96476 non-null object
order_estimated_delivery_date    99441 non-null object
review_id                        98926 non-null object
review_score                     98926 non-null float64
review_comment_title             11659 non-null object
review_comment_message           41343 non-null object
review_creation_date             98926 non-null object
review_answer_timestamp          98926 non-null object
dtypes: float64(1), object(13)
memory usage: 11.4+ MB


In [68]:
# Deleção dos campos não necessários (manter somente os marcados pelo retângulo verde na figura acima)
orders_reviews = orders_reviews_apos_merge.drop(['review_comment_title', 'review_comment_message','review_creation_date', 'review_answer_timestamp'], axis=1)

In [69]:
orders.shape, reviews_normalizada.shape, orders_reviews_apos_merge.shape, orders_reviews.shape

((99441, 8), (98926, 7), (99441, 14), (99441, 10))

In [70]:
orders_reviews_apos_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99441 entries, 0 to 99440
Data columns (total 14 columns):
order_id                         99441 non-null object
customer_id                      99441 non-null object
order_status                     99441 non-null object
order_purchase_timestamp         99441 non-null object
order_approved_at                99281 non-null object
order_delivered_carrier_date     97658 non-null object
order_delivered_customer_date    96476 non-null object
order_estimated_delivery_date    99441 non-null object
review_id                        98926 non-null object
review_score                     98926 non-null float64
review_comment_title             11659 non-null object
review_comment_message           41343 non-null object
review_creation_date             98926 non-null object
review_answer_timestamp          98926 non-null object
dtypes: float64(1), object(13)
memory usage: 11.4+ MB


### 4.2 - Merge do resultado com <font color=red>"payments_normalizada"</font> 

In [71]:
orders_reviews_payments = pd.merge(orders_reviews, payments_normalizada, how = 'left', left_on = 'order_id', right_on = 'order_id')

In [72]:
orders_reviews.shape, payments_normalizada.shape, orders_reviews_payments.shape

((99441, 10), (99440, 8), (99441, 17))

##### Há um 'order_id' a menos na tabela <font color=red>"payments_normalizada"</font>.
> Os 07 (sete) campos juntados devem apresentar 'NaN'.

> Tenho curiosidade de descobrir qual registro é esse.

#### 4.2.1 -Investigação sobre o 'order_id' sem correspondência
> vou selecionar campo a campo, os registros que são 'nan'.

In [73]:
orders_reviews_payments.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99441 entries, 0 to 99440
Data columns (total 17 columns):
order_id                         99441 non-null object
customer_id                      99441 non-null object
order_status                     99441 non-null object
order_purchase_timestamp         99441 non-null object
order_approved_at                99281 non-null object
order_delivered_carrier_date     97658 non-null object
order_delivered_customer_date    96476 non-null object
order_estimated_delivery_date    99441 non-null object
review_id                        98926 non-null object
review_score                     98926 non-null float64
qtde_boleto                      19784 non-null float64
qtde_credit_card                 76505 non-null float64
qtde_debit_card                  1528 non-null float64
qtde_not_defined                 3 non-null float64
qtde_voucher                     3866 non-null float64
soma_payment                     99440 non-null float64
qtde_insta

In [74]:
# Cria um dataframe somente com 'order_id' e os campos juntados
ateh_agora = orders_reviews_payments[['order_id', 'qtde_boleto', 'qtde_credit_card', 'qtde_debit_card', 'qtde_not_defined', 'qtde_voucher', 'soma_payment', 'qtde_installments']]

In [75]:
ateh_agora.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99441 entries, 0 to 99440
Data columns (total 8 columns):
order_id             99441 non-null object
qtde_boleto          19784 non-null float64
qtde_credit_card     76505 non-null float64
qtde_debit_card      1528 non-null float64
qtde_not_defined     3 non-null float64
qtde_voucher         3866 non-null float64
soma_payment         99440 non-null float64
qtde_installments    99440 non-null float64
dtypes: float64(7), object(1)
memory usage: 6.8+ MB


In [76]:
ateh_agora.head()

order_id  qtde_boleto  qtde_credit_card  \
0  e481f51cbdc54678b7cc49136f2d6af7          nan              1.00   
1  53cdb2fc8bc7dce0b6741e2150273451         1.00               nan   
2  47770eb9100c2d0c44946d9cf07ec65d          nan              1.00   
3  949d5b44dbf5de918fe9c16f97b45f8a          nan              1.00   
4  ad21c59c0840e6cb83a9ceb5573f8159          nan              1.00   

   qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
0              nan               nan          2.00         38.71   
1              nan               nan           nan        141.46   
2              nan               nan           nan        179.12   
3              nan               nan           nan         72.20   
4              nan               nan           nan         28.62   

   qtde_installments  
0               1.00  
1               1.00  
2               3.00  
3               1.00  
4               1.00

In [77]:
ateh_agora.shape

(99441, 8)

In [78]:
# cria novo dataframe com os registros que têm 'qtde_boleto' como 'nan'
ateh_agora2 = ateh_agora.loc[ateh_agora['qtde_boleto'].isna()].sort_values(by='order_id')

In [79]:
ateh_agora2.head()

order_id  qtde_boleto  qtde_credit_card  \
85267  00010242fe8c5a6d1ba2dd792cb16214          nan              1.00   
71853  00018f77f2f0320c557190d7a144bdd3          nan              1.00   
6298   000229ec398224ef6ca0657da4fc703e          nan              1.00   
22550  00024acbcdf0a6daa1e931b038114c75          nan              1.00   
5247   00042b26cf59d7ce69dfabb4e55b4fd9          nan              1.00   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
85267              nan               nan           nan         72.19   
71853              nan               nan           nan        259.83   
6298               nan               nan           nan        216.87   
22550              nan               nan           nan         25.78   
5247               nan               nan           nan        218.04   

       qtde_installments  
85267               2.00  
71853               3.00  
6298                5.00  
22550               2.00  
5247                3.00

In [80]:
ateh_agora2.shape

(79657, 8)

In [81]:
# Cria novo dataframe com os registros de "ateh_agora2" que têm 'qtde_credit_card' como 'nan'
ateh_agora3 = ateh_agora2.loc[ateh_agora2['qtde_credit_card'].isna()].sort_values(by='order_id')

In [82]:
ateh_agora3.head()

order_id  qtde_boleto  qtde_credit_card  \
2889   002f16b7bc4530031b7d90f791b12d8a          nan               nan   
44054  002f19a65a2ddd70a090297872e6d64e          nan               nan   
9473   0036757472ece3dde52fd4bfd929c90e          nan               nan   
85919  004345d16a1ab2c21962992c721c8643          nan               nan   
22168  0071ee2429bc1efdc43aa3e073a5290e          nan               nan   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
2889              1.00               nan           nan        284.13   
44054              nan               nan          2.00         77.29   
9473              1.00               nan           nan        203.03   
85919              nan               nan          1.00         53.27   
22168              nan               nan          2.00        192.44   

       qtde_installments  
2889                1.00  
44054               1.00  
9473                1.00  
85919               1.00  
22168               1.00

In [83]:
ateh_agora3.shape

(3152, 8)

In [84]:
# Cria novo dataframe com os registros de "ateh_agora3" que têm 'qtde_debit_card' como 'nan'
ateh_agora4 = ateh_agora3.loc[ateh_agora3['qtde_debit_card'].isna()].sort_values(by='order_id')

In [85]:
ateh_agora4.head()

order_id  qtde_boleto  qtde_credit_card  \
44054  002f19a65a2ddd70a090297872e6d64e          nan               nan   
85919  004345d16a1ab2c21962992c721c8643          nan               nan   
22168  0071ee2429bc1efdc43aa3e073a5290e          nan               nan   
89147  0080eebb288dba1857ccf048dfe6bdfe          nan               nan   
6995   00995d799817ecc3bd2abd8fbe59c430          nan               nan   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
44054              nan               nan          2.00         77.29   
85919              nan               nan          1.00         53.27   
22168              nan               nan          2.00        192.44   
89147              nan               nan          1.00         21.10   
6995               nan               nan          1.00         62.82   

       qtde_installments  
44054               1.00  
85919               1.00  
22168               1.00  
89147               1.00  
6995                1.00

In [86]:
ateh_agora4.shape

(1625, 8)

In [87]:
# Mesmo processo com o próximo campo que é 'qtde_not_defined'
ateh_agora5 = ateh_agora4.loc[ateh_agora4['qtde_not_defined'].isna()].sort_values(by='order_id')

In [88]:
ateh_agora5.head()

order_id  qtde_boleto  qtde_credit_card  \
44054  002f19a65a2ddd70a090297872e6d64e          nan               nan   
85919  004345d16a1ab2c21962992c721c8643          nan               nan   
22168  0071ee2429bc1efdc43aa3e073a5290e          nan               nan   
89147  0080eebb288dba1857ccf048dfe6bdfe          nan               nan   
6995   00995d799817ecc3bd2abd8fbe59c430          nan               nan   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
44054              nan               nan          2.00         77.29   
85919              nan               nan          1.00         53.27   
22168              nan               nan          2.00        192.44   
89147              nan               nan          1.00         21.10   
6995               nan               nan          1.00         62.82   

       qtde_installments  
44054               1.00  
85919               1.00  
22168               1.00  
89147               1.00  
6995                1.00

In [89]:
ateh_agora5.shape # o número de registros vai diminuindo

(1622, 8)

In [90]:
# Mais um campo: 'qtde_voucher'
ateh_agora6 = ateh_agora5.loc[ateh_agora5['qtde_voucher'].isna()].sort_values(by='order_id')

In [91]:
ateh_agora6.head()

order_id  qtde_boleto  qtde_credit_card  \
30710  bfbd0f9bdef84302105ad712db648a6c          nan               nan   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
30710              nan               nan           nan           nan   

       qtde_installments  
30710                nan

In [92]:
ateh_agora6.shape

(1, 8)

###### Pronto: o registro acima já é o que procuramos, todos os campos importados são 'nan'
> Mas vou continuar o processo até varrer todos os campos.

> São só mais dois campos.

In [93]:
ateh_agora7 = ateh_agora6.loc[ateh_agora6['soma_payment'].isna()].sort_values(by='order_id')

In [94]:
ateh_agora7.head()

order_id  qtde_boleto  qtde_credit_card  \
30710  bfbd0f9bdef84302105ad712db648a6c          nan               nan   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
30710              nan               nan           nan           nan   

       qtde_installments  
30710                nan

In [95]:
ateh_agora7.shape

(1, 8)

In [96]:
ateh_agora8 = ateh_agora7.loc[ateh_agora7['qtde_installments'].isna()].sort_values(by='order_id')

In [97]:
ateh_agora8.head()

order_id  qtde_boleto  qtde_credit_card  \
30710  bfbd0f9bdef84302105ad712db648a6c          nan               nan   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
30710              nan               nan           nan           nan   

       qtde_installments  
30710                nan

In [98]:
ateh_agora8.shape

(1, 8)

In [99]:
ateh_agora8

order_id  qtde_boleto  qtde_credit_card  \
30710  bfbd0f9bdef84302105ad712db648a6c          nan               nan   

       qtde_debit_card  qtde_not_defined  qtde_voucher  soma_payment  \
30710              nan               nan           nan           nan   

       qtde_installments  
30710                nan

###### Vamos ver qual order é essa.

In [100]:
# Listagem do 'order_id' que não tem payment correspondente.
orders_reviews_payments.loc[orders_reviews_payments['order_id'] == "bfbd0f9bdef84302105ad712db648a6c"].sort_values(by='order_id')

order_id                       customer_id  \
30710  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   

      order_status order_purchase_timestamp    order_approved_at  \
30710    delivered      2016-09-15 12:16:38  2016-09-15 12:16:38   

      order_delivered_carrier_date order_delivered_customer_date  \
30710          2016-11-07 17:11:53           2016-11-09 07:47:38   

      order_estimated_delivery_date                         review_id  \
30710           2016-10-04 00:00:00  6916ca4502d6d3bfd39818759d55d536   

       review_score  qtde_boleto  qtde_credit_card  qtde_debit_card  \
30710          1.00          nan               nan              nan   

       qtde_not_defined  qtde_voucher  soma_payment  qtde_installments  
30710               nan           nan           nan                nan

###### Não identifiquei nenhuma característica que pudesse explicar a ausência de correspondência desse valor na tabela <font color=red>"payments_normalizada"</font> 
> Alguém aí poderia passar um Twit pra mim, explicando?

> Se não receber nada, vou acabar deletando este registro.
>> Vocês 72 horas, a partir de 26/04/2021!!!

### 4.3 - Merge do resultado com <font color=red>"customers"</font> 

In [101]:
orders_reviews_payments_customers = pd.merge(orders_reviews_payments, customers, how = 'left', on = 'customer_id')

In [102]:
orders_reviews_payments.shape, customers.shape, orders_reviews_payments_customers.shape

((99441, 17), (99441, 5), (99441, 21))

### 4.4 - Merge do resultado até agora com <font color=red>"itens_sem_repetir"</font> 

###### A tabela <font color=red>"itens_sem_repetir"</font> tem 98,6 mil registros, enquanto a tabela <font color=red>"orders_reviews_payments_customers"</font> tem 99,4 mil.

In [103]:
itens_sem_repetir.shape, orders_reviews_payments_customers.shape

((98666, 7), (99441, 21))

#### <font color=magenta>CONSIDERAÇÕES</font>.
> <font color=red>'itens_sem_repetir'</font> fornece o produto e o vendedor de cada compra. 
>> Deveria haver correspondência exata entre cada 'order_id' dela com os da tabela <font color=red>"orders"</font>

> Do jeito que as duas tabelas estão, parece que há compras sem item algum. 
>> É uma incongruência.

In [104]:
Image (url = "Detalhe orders e itens_sem_repetir.PNG")

###### Efetuaremos o merge da tabela <font color=red>"itens_sem_repetir"</font> , mas, no tabelão, teremos de tratar esses 'valores nulos' que surgirão.

In [105]:
ord_rev_pay_cus_itens_apos_merge = pd.merge(orders_reviews_payments_customers, itens_sem_repetir, how = 'left', on = 'order_id')

In [106]:
# Deleção dos campos não necessários: 'order_item_id', 'shipping_limit_date', 'price' e 'freight_value'
ord_rev_pay_cus_itens = ord_rev_pay_cus_itens_apos_merge.drop(['order_item_id', 'shipping_limit_date','price', 'freight_value'], axis=1)

In [107]:
orders_reviews_payments_customers.shape, itens_sem_repetir.shape, ord_rev_pay_cus_itens_apos_merge.shape, ord_rev_pay_cus_itens.shape

((99441, 21), (98666, 7), (99441, 27), (99441, 23))

### 4.5 - Merge do resultado até agora com <font color=red>"products"</font> 

In [108]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
product_id                    32951 non-null object
product_category_name         32341 non-null object
product_name_lenght           32341 non-null float64
product_description_lenght    32341 non-null float64
product_photos_qty            32341 non-null float64
product_weight_g              32949 non-null float64
product_length_cm             32949 non-null float64
product_height_cm             32949 non-null float64
product_width_cm              32949 non-null float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [109]:
# Há menos 'product_category_name' do que 'product_Id', ai ai...
products_sem_category = products.loc[products['product_category_name'].isnull()].sort_values(by='product_id')

In [110]:
products_sem_category.shape

(610, 9)

In [111]:
products_sem_category.head() # Temos 610 'product_id' sem suas categorias

product_id product_category_name  \
6039   0082684bb4a60a862baaf7a60a5845ed                   NaN   
13328  00ab8a8b9fe219511dc3f178c6d79698                   NaN   
6495   00d62b338366db4c4aec8547ea8f928e                   NaN   
5832   0103863bf3441460142ec23c74388e4c                   NaN   
1531   014fcf6bd5cd4c7ee29fb3bb618c445e                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
6039                   nan                         nan                 nan   
13328                  nan                         nan                 nan   
6495                   nan                         nan                 nan   
5832                   nan                         nan                 nan   
1531                   nan                         nan                 nan   

       product_weight_g  product_length_cm  product_height_cm  \
6039             500.00              24.00               4.00   
13328          2,100.00              50.00              30.00   
6495             400.00              24.00               4.00   
5832             200.00              16.00               2.00   
1531           7,000.00              55.00              30.00   

       product_width_cm  
6039              15.00  
13328             30.00  
6495              15.00  
5832              11.00  
1531              45.00

In [112]:
# Merge com "products"
ord_rev_pay_cus_itens_pro_apos_merge = pd.merge(ord_rev_pay_cus_itens, products, how = 'left', on = 'product_id')

In [113]:
# Deleção dos campos não necessários: todos de "products", exceto 'product_id' e 'product_category_name'
ord_rev_pay_cus_itens_pro = ord_rev_pay_cus_itens_pro_apos_merge.drop(['product_name_lenght', 'product_description_lenght','product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm'], axis=1)

In [114]:
ord_rev_pay_cus_itens.shape, products.shape, ord_rev_pay_cus_itens_pro_apos_merge.shape, ord_rev_pay_cus_itens_pro.shape

((99441, 23), (32951, 9), (99441, 31), (99441, 24))

### 4.6 - Merge do resultado até agora com <font color=red>"sellers_normalizado"</font> 

In [115]:
ord_rev_pay_cus_itens_pro_se_pos_merge = pd.merge(ord_rev_pay_cus_itens_pro, sellers_normalizado, how = 'left', on = 'seller_id')

In [116]:
# Deleção do campo não necessário: 'business_segment'
ord_rev_pay_cus_itens_pro_se = ord_rev_pay_cus_itens_pro_se_pos_merge.drop(['business_segment'], axis=1)

In [117]:
ord_rev_pay_cus_itens_pro_se_pos_merge.shape, ord_rev_pay_cus_itens_pro_se.shape

((99441, 28), (99441, 27))

In [118]:
ord_rev_pay_cus_itens_pro_se.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99441 entries, 0 to 99440
Data columns (total 27 columns):
order_id                         99441 non-null object
customer_id                      99441 non-null object
order_status                     99441 non-null object
order_purchase_timestamp         99441 non-null object
order_approved_at                99281 non-null object
order_delivered_carrier_date     97658 non-null object
order_delivered_customer_date    96476 non-null object
order_estimated_delivery_date    99441 non-null object
review_id                        98926 non-null object
review_score                     98926 non-null float64
qtde_boleto                      19784 non-null float64
qtde_credit_card                 76505 non-null float64
qtde_debit_card                  1528 non-null float64
qtde_not_defined                 3 non-null float64
qtde_voucher                     3866 non-null float64
soma_payment                     99440 non-null float64
qtde_insta

##### Optaremos por deixar as coordenadas para um momento posterior.
> Nesse momento, temos uma tabela única com 99.441 registros e 27 campos.
>> Trabalho nesse tabela deverá ser executado sobre valores nulos, NaN, etc.

In [119]:
ord_rev_pay_cus_itens_pro_se.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score  qtde_boleto  qtde_credit_card  qtde_debit_card  \
0          4.00          nan              1.00              nan   
1          4.00         1.00               nan              nan   
2          5.00          nan              1.00              nan   
3          5.00          nan              1.00              nan   
4          5.00          nan              1.00              nan   

   qtde_not_defined  qtde_voucher  soma_payment  qtde_installments  \
0               nan          2.00         38.71               1.00   
1               nan           nan        141.46               1.00   
2               nan           nan        179.12               3.00   
3               nan           nan         72.20               1.00   
4               nan           nan         28.62               1.00   

                 customer_unique_id  customer_zip_code_prefix  \
0  7c396fd4830fd04220f754e42b4e5bff                      3149   
1  af07308b275d755c9edb36a90c618231                     47813   
2  3a653a41f6f9fc3d2a113cf8398680e8                     75265   
3  7c142cf63193a1473d2e66489a9ae977                     59296   
4  72632f0f9dd73dfee390c9b22eb56dd6                      9195   

             customer_city customer_state                        product_id  \
0                sao paulo             SP  87285b34884572647811a353c7ac498a   
1                barreiras             BA  595fac2a385ac33a80bd5114aec74eb8   
2               vianopolis             GO  aa4383b373c6aca5d8797843e5594415   
3  sao goncalo do amarante             RN  d0b61bfb1de832b15ba9d266ca96e5b0   
4              santo andre             SP  65266b2da20d04dbe00c5c2d3bb7859e   

                          seller_id  product_category_name  \
0  3504c0cb71d7fa48d967e0e4c94d59d9  utilidades_domesticas   
1  289cdb325fb7e7f891c38608bf9e0962             perfumaria   
2  4869f7a5dfa277a7dca6462dcf3b52b2             automotivo   
3  66922902710d126a0e7d26b0e3805106               pet_shop   
4  2c9e548be18521d1c43cde1c582c6de8              papelaria   

   seller_zip_code_prefix      seller_city seller_state  
0                9,350.00             maua           SP  
1               31,570.00   belo horizonte           SP  
2               14,840.00          guariba           SP  
3               31,842.00   belo horizonte           MG  
4                8,752.00  mogi das cruzes        

In [120]:
ord_rev_pay_cus_itens_pro_se_ordenado = ord_rev_pay_cus_itens_pro_se.sort_values(by='customer_city')

In [121]:
ord_rev_pay_cus_itens_pro_se_ordenado.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 99441 entries, 49914 to 83980
Data columns (total 27 columns):
order_id                         99441 non-null object
customer_id                      99441 non-null object
order_status                     99441 non-null object
order_purchase_timestamp         99441 non-null object
order_approved_at                99281 non-null object
order_delivered_carrier_date     97658 non-null object
order_delivered_customer_date    96476 non-null object
order_estimated_delivery_date    99441 non-null object
review_id                        98926 non-null object
review_score                     98926 non-null float64
qtde_boleto                      19784 non-null float64
qtde_credit_card                 76505 non-null float64
qtde_debit_card                  1528 non-null float64
qtde_not_defined                 3 non-null float64
qtde_voucher                     3866 non-null float64
soma_payment                     99440 non-null float64
qtde_i

In [122]:
# ord_rev_pay_cus_itens_pro_se_ordenado.to_excel("ord_rev_pay_cus_itens_pro_se_ordenado.xlsx")